lambda/transl_array_comprehension.ml

```diff
@@ -0,0 +1,932 @@
+    lambda] function awaiting the body of the loop or conditional corresponding
+    to this clause.  The argument to that function will be executed once for
+    every tuple of elements being iterated over in the [for ... and ...] case,
+    or it will be executed iff the condition is true in the [when] case.
+
+    This function is only used if we are not in the fixed-size array case; see
+    [clauses] and [for_and_clause] for more details. *)
+let clause ~transl_exp ~scopes ~loc = function
+  | Texp_comp_for bindings ->
+      let make_clause, var_bindings =
+        for_and_clause ~transl_exp ~loc ~scopes ~usage:Once bindings
+      in
+      fun body -> Let_binding.let_all
+                    (Iterator_bindings.all_let_bindings var_bindings)
+                    (make_clause body)
+  | Texp_comp_when cond ->
+      fun body -> Lifthenelse(transl_exp ~scopes cond,
+                    body,
+                    lambda_unit,
+                    Pintval (* [unit] is immediate *))
+
+(** The [array_sizing] type describes whether an array comprehension has been
+    translated using the fixed-size array optimization ([Fixed_size]), or it has
+    not been but instead been translated using the usual dynamically-sized array
+    ([Dynamic_size]).
+
+    If an array comprehension is of the form
+    {[
+      [|BODY for ITER and ITER ... and ITER|]
+    ]}
+    then we can compute the size of the resulting array before allocating it
+    ([Fixed_size]); otherwise, we cannot ([Dynamic_size]), and we have to
+    dynamically grow the array as we iterate and shrink it to size at the
+    end. *)
+type array_sizing =
+  | Fixed_size
+  | Dynamic_size
+
+(** The [array_size] type provides both the variable holding the array size
+    ([array_size]) and a description of how the array size has been/is being
+    computed ([array_sizing]; see the [array_sizing] type for more details).  In
+    the case where the array has been translated with the fixed-size array
+    optimization (when [sizing] is [Fixed_size]), the variable holding the size
+    is immutable; in the usual dynamically-sized array case (when [sizing] is
+    [Dynamic_size]), the variable holding the size is be mutable so that the
+    array size can be queried and grown. *)
+type array_size =
+  { array_size   : Ident.t
+  ; array_sizing : array_sizing
+  }
+(* CR aspectorzabusky: The names [array_size] (this type), [array_size] (the
+   field name), and [array_sizing] are not great, but I can't think of anything
+   better. *)
+
+(** The result of translating the clauses portion of an array comprehension
+    (everything but the body) *)
+type translated_clauses =
+  { array_size         : array_size
+  (** Whether the array is of a fixed size or must be grown dynamically, and the
+      attendant information; see the [array_size] type for more details. *)
+  ; outside_context    : lambda -> lambda
+  (** The context that must be in force throughout the entire translated
```

Comment:
No, it's for the fixed-size array case (of course that's where the complexity comes from).  I replaced the last sentence with the following: "This is used in the fixed-size array case to collect the lengths of the iterators being iterated over, which determine the size of the array."  Does that help?

Reply:
I think the problem is that I don't know what "context that must be in force" means. And how that phrase relates to the fact that the type of this field is `lambda -> lambda`. Maybe this is an expression context? But that use of context is rather rare, even for a pointy-headed PL researcher. (My first meaning for "context" is some dictionary-like structure.) I think it might be better to rephrase this as a function that wraps some other lambda term.

Reply:
These confusions are all very reasonable, thanks for helping clarify this.  I think it might be a sign that the design is a bit messy here; this is a function that's either "do some complicated setup work for the fixed-size array case" or "`id`".  The thing is that we learn about the complicated setup work that's necessary when we translate the clauses, but _applying_ that complicated setup work has to be the very last thing we do because it comes first.  So it's natural -- at least, it felt natural to me -- to make this a function that does the setup work.  When not `id`, we're in the fixed-size array case and the function is roughly

```
fun EXP ->
  let (* ... all iterator bindings (start, stop, arrays, lengths) ... *) = (* ... all iterators ... *) in
  if (* any iterator is empty *) then
    [||]
  else
    EXP
```

So the `outside_context` is indeed an expression context, it's more complex than just a list of bindings, and it needs to come before the stuff that's common to every comprehension.  I could attempt to further clarify this with comments and naming, or I could try an alternative design.

Reply:
I favor an alternative design here. This goes with "make illegal states unrepresentable". Something of type `lambda -> lambda` has a lot of power! But a `value_binding option` (perhaps? haven't checked the types) gets right to the point. The type doesn't directly say that there should be a conditional, too, but whatever consumes this will add that bit. I just think this will be easier to work with (to read/understand and to debug).

Reply:
I made this change in commit bc61dda8cfa61cf755d2050029a9f89625488515, and I do think it's better.  It also involved some other changes to the types there: `array_sizing` is gone, and `array_sizing` got a friend named `array_sizing_info`.  Probably worth taking a look at, but yes, the alternative design in broad strokes is clearly better!